[Issue] Linux gfx90a Pytorch smoketests crash: `HIP_VISIBLE_DEVICES` + `GPU_DEVICE_ORDINAL` conflict → `agent.cpp:245 FATAL` (linux-gfx90a-gpu-rocm runners)

## Summary

  The "Run PyTorch smoketests" step in the `multi_arch_release_linux_pytorch_wheels.yml` workflow crashes immediately on `linux-gfx90a-gpu-rocm` runners before any tests execute. The HIP runtime issues a fatal `F`-level abort in `agent.cpp:245` because the runner's GPU isolation env-file sets both `HIP_VISIBLE_DEVICES=0` and `GPU_DEVICE_ORDINAL` to a different ordinal (1, 2, or 4 depending on the runner) simultaneously —conflicting values that the HIP runtime treats as a fatal error.

  ## Context

  - **Workflow:** `multi_arch_release_linux_pytorch_wheels.yml` (ROCm/rockrel)
  - **Failing run:** [↗ Run 27806073341](https://github.com/ROCm/rockrel/actions/runs/27806073341)
  - **Failing job:** [↗ Test PyTorch | gfx90a (job 82314886472)](https://github.com/ROCm/rockrel/actions/runs/27806073341/job/82314886472)
  - **Affected runners:** multiple (`smci250-ccs-aus-c05-33`, `smci250-ccs-aus-c05-13`, `smci250-ccs-aus-c03-38`) (pool: `linux-gfx90a-gpu-rocm`)
  - **Affected arch:** `gfx90a`
  - **Step:** "Run PyTorch smoketests"

  ## Failing Jobs (torch release/2.12, all Python versions)

  - [py 3.10 — job 82314886472](https://github.com/ROCm/rockrel/actions/runs/27806073341/job/82314886472)
  - [py 3.11 — job 82315777463](https://github.com/ROCm/rockrel/actions/runs/27806073341/job/82315777463)
  - [py 3.12 — job 82315706085](https://github.com/ROCm/rockrel/actions/runs/27806073341/job/82315706085)
  - [py 3.13 — job 82315774119](https://github.com/ROCm/rockrel/actions/runs/27806073341/job/82315774119)
  - [py 3.14 — job 82315852757](https://github.com/ROCm/rockrel/actions/runs/27806073341/job/82315852757)

  ## Error

  ```
  HIP_VISIBLE_DEVICES already set to: 0
  Getting GPU information from PyTorch...
  [ERROR] Failed to retrieve GPU info: E-001h rocSHMEM Could not open libnuma. Returning    NUMAWrapper@src/gda/numa_wrapper.cpp:48
  F0619 08:45:09.227348     958 agent.cpp:245] Conflicting visibility of agent-2 between HIP_VISIBLE_DEVICES and GPU_DEVICE_ORDINAL. Assuming
  HIP_VISIBLE_DEVICES supersedes GPU_DEVICE_ORDINAL
  *** Check failure stack trace: ***
  ##[error]Process completed with exit code 1.
  ```

  ## Suspected Root Cause

  The container is started (from `test_pytorch_wheels.yml`) with the runner's env-file (`/etc/podinfo/gha-gpu-isolation-settings`) injecting both:

  ```
  HIP_VISIBLE_DEVICES=0
  GPU_DEVICE_ORDINAL=1   # varies per runner: 1, 2, or 4
  ROCR_VISIBLE_DEVICES=1 # matches GPU_DEVICE_ORDINAL, not HIP_VISIBLE_DEVICES
  ```

  These two variables refer to different GPU ordinals, and the HIP runtime's `agent.cpp:245` detects the conflict and issues a fatal abort. The `rocSHMEM libnuma` message is an unrelated warning on stderr (tracked in #5459); the actual killer is the `F0619 … agent.cpp:245` line.

  The code path is: `run_pytorch_smoke_tests.py` → `get_unique_supported_devices()` → `get_supported_and_visible_gpus()` → `subprocess.run([python,-c, "import torch; torch.cuda.device_count()…"])` → HIP fatal crash → `CalledProcessError` → `sys.exit(1)`.

--- a/external-builds/pytorch/pytorch_utils.py
+++ b/external-builds/pytorch/pytorch_utils.py
@@ -18,6 +18,9 @@ def reconcile_agent_visibility_env() -> None:
     The runner scripts always set HIP_VISIBLE_DEVICES, which supersedes and
     conflicts with a GPU_DEVICE_ORDINAL injected by the CI runner's isolation
     env-file (HIP aborts fatally on the mismatch).
+
+    See https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html for
+    multiple ways to achieve isolation of GPUs in the ROCm software stack.
     """
     gpu_device_ordinal = os.environ.pop("GPU_DEVICE_ORDINAL", None)
     if gpu_device_ordinal is not None:

--- a/external-builds/pytorch/pytorch_utils.py
+++ b/external-builds/pytorch/pytorch_utils.py
@@ -46,10 +46,6 @@ def get_supported_and_visible_gpus() -> tuple[list[str], list[str]]:
             - visible_gpus: List of AMDGPU archs physically visible
         Exits on failure.
     """
-    # Runs before the first torch subprocess, so any conflicting
-    # GPU_DEVICE_ORDINAL is removed before HIP initializes.
-    reconcile_agent_visibility_env()
-
     query_script = """
 import sys
 try:

--- a/external-builds/pytorch/pytorch_utils.py
+++ b/external-builds/pytorch/pytorch_utils.py
@@ -1,7 +1,13 @@
 # Copyright Advanced Micro Devices, Inc.
 # SPDX-License-Identifier: MIT
 
-"""Shared utilities for PyTorch testing."""
+"""Shared utilities for PyTorch testing.
+
+Several helpers here manage GPU visibility/isolation through HIP_VISIBLE_DEVICES,
+ROCR_VISIBLE_DEVICES, and GPU_DEVICE_ORDINAL. See
+https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html for the
+multiple ways to achieve isolation of GPUs in the ROCm software stack.
+"""
 
 import os
 import subprocess
@@ -18,9 +24,6 @@ def reconcile_agent_visibility_env() -> None:
     The runner scripts always set HIP_VISIBLE_DEVICES, which supersedes and
     conflicts with a GPU_DEVICE_ORDINAL injected by the CI runner's isolation
     env-file (HIP aborts fatally on the mismatch).
-
-    See https://rocm.docs.amd.com/en/latest/conceptual/gpu-isolation.html for
-    multiple ways to achieve isolation of GPUs in the ROCm software stack.
     """
     gpu_device_ordinal = os.environ.pop("GPU_DEVICE_ORDINAL", None)
     if gpu_device_ordinal is not None:

--- a/external-builds/pytorch/pytorch_utils.py
+++ b/external-builds/pytorch/pytorch_utils.py
@@ -13,13 +13,13 @@ from packaging.version import Version
 
 
 def reconcile_agent_visibility_env() -> None:
-    """Drop GPU_DEVICE_ORDINAL when HIP_VISIBLE_DEVICES is set.
+    """Drop GPU_DEVICE_ORDINAL to avoid a fatal HIP agent-visibility conflict.
 
-    Avoids a fatal HIP agent-visibility conflict when a runner sets both;
-    HIP_VISIBLE_DEVICES supersedes.
+    The runner scripts always set HIP_VISIBLE_DEVICES, which supersedes and
+    conflicts with a GPU_DEVICE_ORDINAL injected by the CI runner's isolation
+    env-file (HIP aborts fatally on the mismatch).
     """
-    if os.environ.get("HIP_VISIBLE_DEVICES"):
-        os.environ.pop("GPU_DEVICE_ORDINAL", None)
+    os.environ.pop("GPU_DEVICE_ORDINAL", None)
 
 
 def get_supported_and_visible_gpus() -> tuple[list[str], list[str]]:
@@ -41,6 +41,10 @@ def get_supported_and_visible_gpus() -> tuple[list[str], list[str]]:
             - visible_gpus: List of AMDGPU archs physically visible
         Exits on failure.
     """
+    # Runs before the first torch subprocess, so any conflicting
+    # GPU_DEVICE_ORDINAL is removed before HIP initializes.
+    reconcile_agent_visibility_env()
+
     query_script = """
 import sys
 try:

--- a/external-builds/pytorch/pytorch_utils.py
+++ b/external-builds/pytorch/pytorch_utils.py
@@ -19,7 +19,12 @@ def reconcile_agent_visibility_env() -> None:
     conflicts with a GPU_DEVICE_ORDINAL injected by the CI runner's isolation
     env-file (HIP aborts fatally on the mismatch).
     """
-    os.environ.pop("GPU_DEVICE_ORDINAL", None)
+    gpu_device_ordinal = os.environ.pop("GPU_DEVICE_ORDINAL", None)
+    if gpu_device_ordinal is not None:
+        print(
+            f"Warning: dropped conflicting GPU_DEVICE_ORDINAL={gpu_device_ordinal} "
+            "(HIP_VISIBLE_DEVICES supersedes it)"
+        )
 
 
 def get_supported_and_visible_gpus() -> tuple[list[str], list[str]]:

--- a/external-builds/pytorch/pytorch_utils.py
+++ b/external-builds/pytorch/pytorch_utils.py
@@ -12,6 +12,16 @@ from importlib.metadata import version as get_package_version
 from packaging.version import Version
 
 
+def reconcile_agent_visibility_env() -> None:
+    """Drop GPU_DEVICE_ORDINAL when HIP_VISIBLE_DEVICES is set.
+
+    Avoids a fatal HIP agent-visibility conflict when a runner sets both;
+    HIP_VISIBLE_DEVICES supersedes.
+    """
+    if os.environ.get("HIP_VISIBLE_DEVICES"):
+        os.environ.pop("GPU_DEVICE_ORDINAL", None)
+
+
 def get_supported_and_visible_gpus() -> tuple[list[str], list[str]]:
     """Get both supported and visible GPUs in a single subprocess call.
 

--- a/external-builds/pytorch/pytorch_utils.py
+++ b/external-builds/pytorch/pytorch_utils.py
@@ -22,8 +22,8 @@ def reconcile_agent_visibility_env() -> None:
     gpu_device_ordinal = os.environ.pop("GPU_DEVICE_ORDINAL", None)
     if gpu_device_ordinal is not None:
         print(
-            f"Warning: dropped conflicting GPU_DEVICE_ORDINAL={gpu_device_ordinal} "
-            "(HIP_VISIBLE_DEVICES supersedes it)"
+            f"[WARNING] Unset GPU_DEVICE_ORDINAL={gpu_device_ordinal} "
+            "(HIP_VISIBLE_DEVICES takes precedence)"
         )
 
 

--- a/external-builds/pytorch/run_pytorch_smoke_tests.py
+++ b/external-builds/pytorch/run_pytorch_smoke_tests.py
@@ -27,10 +27,16 @@ from pathlib import Path
 
 from pytorch_utils import (
     get_unique_supported_devices,
+    reconcile_agent_visibility_env,
     set_gpu_execution_policy,
 )
 
 THIS_SCRIPT_DIR = Path(__file__).resolve().parent
+
+
+def setup_env() -> None:
+    """Set up environment variables required for the smoke tests."""
+    reconcile_agent_visibility_env()
 
 
 def cmd_arguments(argv: list[str]) -> tuple[argparse.Namespace, list[str]]:
@@ -88,6 +94,8 @@ def main() -> int:
         if not smoke_tests_dir.exists():
             print(f"ERROR: Directory at '{smoke_tests_dir}' does not exist.")
             sys.exit(1)
+
+        setup_env()
 
         # CRITICAL: Query unique devices for iteration.
         # HIP_VISIBLE_DEVICES will be set inside set_gpu_execution_policy.

--- a/external-builds/pytorch/run_pytorch_smoke_tests.py
+++ b/external-builds/pytorch/run_pytorch_smoke_tests.py
@@ -27,6 +27,7 @@ from pathlib import Path
 
 from pytorch_utils import (
     get_unique_supported_devices,
+    reconcile_agent_visibility_env,
     set_gpu_execution_policy,
 )
 
@@ -91,6 +92,7 @@ def main() -> int:
 
         # CRITICAL: Query unique devices for iteration.
         # HIP_VISIBLE_DEVICES will be set inside set_gpu_execution_policy.
+        reconcile_agent_visibility_env()
         unique_devices = get_unique_supported_devices(args.amdgpu_family)
 
         print(f"Will run smoke tests on {len(unique_devices)} unique device(s)")

--- a/external-builds/pytorch/run_pytorch_smoke_tests.py
+++ b/external-builds/pytorch/run_pytorch_smoke_tests.py
@@ -27,7 +27,6 @@ from pathlib import Path
 
 from pytorch_utils import (
     get_unique_supported_devices,
-    reconcile_agent_visibility_env,
     set_gpu_execution_policy,
 )
 
@@ -92,7 +91,6 @@ def main() -> int:
 
         # CRITICAL: Query unique devices for iteration.
         # HIP_VISIBLE_DEVICES will be set inside set_gpu_execution_policy.
-        reconcile_agent_visibility_env()
         unique_devices = get_unique_supported_devices(args.amdgpu_family)
 
         print(f"Will run smoke tests on {len(unique_devices)} unique device(s)")

--- a/external-builds/pytorch/run_pytorch_tests.py
+++ b/external-builds/pytorch/run_pytorch_tests.py
@@ -84,6 +84,7 @@ from pytorch_utils import (
     check_pytorch_source_version,
     configure_gpu_visibility,
     detect_pytorch_version,
+    reconcile_agent_visibility_env,
 )
 
 THIS_SCRIPT_DIR = Path(__file__).resolve().parent
@@ -100,6 +101,8 @@ def setup_env(pytorch_dir: str) -> None:
         - Creates a temporary directory for MIOpen cache
         - Modifies sys.path to include the test directory
     """
+    reconcile_agent_visibility_env()
+
     os.environ["PYTORCH_PRINT_REPRO_ON_FAILURE"] = "0"
     os.environ["PYTORCH_TEST_WITH_ROCM"] = "1"
     os.environ["MIOPEN_CUSTOM_CACHE_DIR"] = tempfile.mkdtemp()
@@ -255,6 +258,8 @@ def main() -> int:
             pytorch_dir=pytorch_dir, allow_mismatch=args.allow_version_mismatch
         )
 
+        setup_env(pytorch_dir)
+
         # CRITICAL: Determine AMDGPU family and set HIP_VISIBLE_DEVICES
         # BEFORE importing torch/running pytest. Once torch.cuda is initialized,
         # changing HIP_VISIBLE_DEVICES has no effect.
@@ -279,8 +284,6 @@ def main() -> int:
         # Allow manual override of test selection
         if args.k:
             tests_to_skip = args.k
-
-        setup_env(pytorch_dir)
 
         pytest_args = [
             f"{pytorch_dir}/test/test_nn.py",

--- a/external-builds/pytorch/run_pytorch_tests.py
+++ b/external-builds/pytorch/run_pytorch_tests.py
@@ -84,6 +84,7 @@ from pytorch_utils import (
     check_pytorch_source_version,
     configure_gpu_visibility,
     detect_pytorch_version,
+    reconcile_agent_visibility_env,
 )
 
 THIS_SCRIPT_DIR = Path(__file__).resolve().parent
@@ -258,6 +259,7 @@ def main() -> int:
         # CRITICAL: Determine AMDGPU family and set HIP_VISIBLE_DEVICES
         # BEFORE importing torch/running pytest. Once torch.cuda is initialized,
         # changing HIP_VISIBLE_DEVICES has no effect.
+        reconcile_agent_visibility_env()
         selected_archs = configure_gpu_visibility(
             args.amdgpu_family, args.device_query, args.gpu_policy
         )

--- a/external-builds/pytorch/run_pytorch_tests.py
+++ b/external-builds/pytorch/run_pytorch_tests.py
@@ -84,7 +84,6 @@ from pytorch_utils import (
     check_pytorch_source_version,
     configure_gpu_visibility,
     detect_pytorch_version,
-    reconcile_agent_visibility_env,
 )
 
 THIS_SCRIPT_DIR = Path(__file__).resolve().parent
@@ -259,7 +258,6 @@ def main() -> int:
         # CRITICAL: Determine AMDGPU family and set HIP_VISIBLE_DEVICES
         # BEFORE importing torch/running pytest. Once torch.cuda is initialized,
         # changing HIP_VISIBLE_DEVICES has no effect.
-        reconcile_agent_visibility_env()
         selected_archs = configure_gpu_visibility(
             args.amdgpu_family, args.device_query, args.gpu_policy
         )

--- a/external-builds/pytorch/run_pytorch_tests_full.py
+++ b/external-builds/pytorch/run_pytorch_tests_full.py
@@ -65,6 +65,7 @@ from pytorch_utils import (
     check_pytorch_source_version,
     configure_gpu_visibility,
     detect_pytorch_version,
+    reconcile_agent_visibility_env,
 )
 
 THIS_SCRIPT_DIR = Path(__file__).resolve().parent
@@ -140,6 +141,8 @@ INDUCTOR_UNIT_TESTS = [
 
 
 def setup_env(pytorch_dir: Path, test_config: str, amdgpu_family: str = "") -> None:
+    reconcile_agent_visibility_env()
+
     os.environ.setdefault("CI", "1")
     build_env = AMDGPU_FAMILY_TO_BUILD_ENV.get(
         amdgpu_family, ROCM_BUILD_ENVIRONMENT_DEFAULT
@@ -519,6 +522,12 @@ def main(argv: list[str]) -> int:
         pytorch_dir=args.pytorch_dir, allow_mismatch=args.allow_version_mismatch
     )
 
+    setup_env(
+        pytorch_dir=args.pytorch_dir,
+        test_config=args.test_config,
+        amdgpu_family=args.amdgpu_family,
+    )
+
     # Set HIP_VISIBLE_DEVICES BEFORE importing torch or running pytest. Once
     # torch.cuda is initialized, changing HIP_VISIBLE_DEVICES has no effect.
     selected_archs = configure_gpu_visibility(
@@ -540,11 +549,6 @@ def main(argv: list[str]) -> int:
             create_skip_list=not args.debug,
         )
 
-    setup_env(
-        pytorch_dir=args.pytorch_dir,
-        test_config=args.test_config,
-        amdgpu_family=args.amdgpu_family,
-    )
     print_env()
 
     if args.test_config == "inductor":

--- a/external-builds/pytorch/run_pytorch_tests_full.py
+++ b/external-builds/pytorch/run_pytorch_tests_full.py
@@ -65,6 +65,7 @@ from pytorch_utils import (
     check_pytorch_source_version,
     configure_gpu_visibility,
     detect_pytorch_version,
+    reconcile_agent_visibility_env,
 )
 
 THIS_SCRIPT_DIR = Path(__file__).resolve().parent
@@ -521,6 +522,7 @@ def main(argv: list[str]) -> int:
 
     # Set HIP_VISIBLE_DEVICES BEFORE importing torch or running pytest. Once
     # torch.cuda is initialized, changing HIP_VISIBLE_DEVICES has no effect.
+    reconcile_agent_visibility_env()
     selected_archs = configure_gpu_visibility(
         args.amdgpu_family, args.device_query, args.gpu_policy
     )

--- a/external-builds/pytorch/run_pytorch_tests_full.py
+++ b/external-builds/pytorch/run_pytorch_tests_full.py
@@ -65,7 +65,6 @@ from pytorch_utils import (
     check_pytorch_source_version,
     configure_gpu_visibility,
     detect_pytorch_version,
-    reconcile_agent_visibility_env,
 )
 
 THIS_SCRIPT_DIR = Path(__file__).resolve().parent
@@ -522,7 +521,6 @@ def main(argv: list[str]) -> int:
 
     # Set HIP_VISIBLE_DEVICES BEFORE importing torch or running pytest. Once
     # torch.cuda is initialized, changing HIP_VISIBLE_DEVICES has no effect.
-    reconcile_agent_visibility_env()
     selected_archs = configure_gpu_visibility(
         args.amdgpu_family, args.device_query, args.gpu_policy
     )

